aten/src/ATen/native/CPUBlas.h

```diff
@@ -186,4 +192,441 @@ void copy(int64_t n, const float *x, int64_t incx, float *y, int64_t incy);
+    }
+  }
+
+    static inline bool need_pack(ScalarType dtype) {
+    if (dtype == ScalarType::Half) {
+      static bool fp16_pack = dnnl::get_effective_cpu_isa() >= dnnl::cpu_isa::avx512_core_amx_fp16;
+      return fp16_pack;
+    } else if (dtype == ScalarType::BFloat16) {
+      static bool bf16_pack = dnnl::get_effective_cpu_isa() >= dnnl::cpu_isa::avx512_core_amx;
+      return bf16_pack;
+    } else {
+      return false;
+    }
+  }
+};
+
+#endif
+
+// Batch-reduce GEMM
+// Operates by the following formula:
+// C = alpha * SUM(A[i] x B[i]) + beta * C, i = 0 to batch size
+// A Base pointer to a tensor A.
+// B Base pointer to a tensor B.
+// Byte offsets vector of pairs of tensors A and B offsets for
+//     each batch. The number of batches must coincide with the
+//     `batch_size` value passed at object construction stage.
+// C Pointer to a tensor C (accumulation buffer).
+// scratchpad Pointer to a scratchpad buffer.
+void brgemm(
+    int64_t M,
+    int64_t N,
+    int64_t K,
+    int64_t bs,
+    int64_t ld_a,
+    int64_t ld_b,
+    int64_t ld_c,
+    const float alpha,
+    const float beta,
+    const at::Half* A,
+    const at::Half* B,
+    const std::vector<std::pair<int64_t, int64_t>>& offsets,
+    float* C);
+
+void brgemm(
+    int64_t M,
+    int64_t N,
+    int64_t K,
+    int64_t bs,
+    int64_t ld_a,
+    int64_t ld_b,
+    int64_t ld_c,
+    const float alpha,
+    const float beta,
+    const at::BFloat16* A,
+    const at::BFloat16* B,
+    const std::vector<std::pair<int64_t, int64_t>>& offsets,
+    float* C);
+
+// Batch-reduce GEMM
+// Batch size is 1, which means it is equivalent to gemm
+void brgemm(
+    int64_t M,
+    int64_t N,
+    int64_t K,
+    int64_t ld_a,
+    int64_t ld_b,
+    int64_t ld_c,
+    const float alpha,
+    const float beta,
+    const at::Half* A,
+    const at::Half* B,
+    float* C);
+
+void brgemm(
+    int64_t M,
+    int64_t N,
+    int64_t K,
+    int64_t ld_a,
+    int64_t ld_b,
+    int64_t ld_c,
+    const float alpha,
+    const float beta,
+    const at::BFloat16* A,
+    const at::BFloat16* B,
+    float* C);
+
+// Release brgemm hardware context
+void brgemm_release();
+
+// Pack B matrix to get better performance if needed
+void pack(
+    int64_t K,
+    int64_t N,
+    int64_t ld_in,
+    int64_t ld_out,
+    ScalarType dt_in,
+    ScalarType dt_out,
+    const void* in,
+    void* out);
+
+// Whether pack is needed in the platform.
+bool need_pack(ScalarType dt_in);
+
+} // namespace at::native::cpublas
+
+#if AT_MKLDNN_ENABLED() && (defined(__x86_64__) || (defined(_M_X64) && !defined(_M_ARM64EC)))
+namespace std {
+template <>
+struct hash<at::native::cpublas::BrgemmKey> {
+  std::size_t operator()(const at::native::cpublas::BrgemmKey& key) const {
+    std::size_t h = std::hash<bool>()(key.beta + 1);
+    h = std::hash<int64_t>()(key.M) ^ (h << 1);
+    h = std::hash<int64_t>()(key.N) ^ (h << 1);
+    h = std::hash<int64_t>()(key.K) ^ (h << 1);
```

Comment:
I guess this worth a note here why we use a subset of the fields to compute the hash key here.